Pilot segfaults when given invalid remote cluster credentials

**Describe the bug**
After installing Istio 0.8 on a fresh GKE version 1.9.7-gke.3 cluster, the `istio-pilot` pod's `discovery` container is repeatedly crashing:

```
$ kubectl logs -n istio-system istio-pilot-6c5c6b586c-6k2tq discovery
2018-06-20T20:36:58.338514Z	warn	Neither --kubeconfig nor --master was specified.  Using the inClusterConfig.  This might not work.
2018-06-20T20:36:58.340485Z	info	clusters configuration (*clusterregistry.ClusterStore)(0xc420275e80)({
 rc: (map[clusterregistry.Metadata]*clusterregistry.RemoteCluster) {
 },
 storeLock: (sync.RWMutex) {
  w: (sync.Mutex) {
   state: (int32) 0,
   sema: (uint32) 0
  },
  writerSem: (uint32) 0,
  readerSem: (uint32) 0,
  readerCount: (int32) 0,
  readerWait: (int32) 0
 }
})

2018-06-20T20:36:58.342244Z	info	mesh configuration (*v1alpha1.MeshConfig)(0xc4204d1740)(mixer_check_server:"istio-policy.istio-system.svc.cluster.local:15004" mixer_report_server:"istio-telemetry.istio-system.svc.cluster.local:15004" proxy_listen_port:15001 connect_timeout:<seconds:1 > ingress_class:"istio" ingress_service:"istio-ingress" ingress_controller_mode:STRICT rds_refresh_delay:<seconds:10 > enable_tracing:true access_log_file:"/dev/stdout" default_config:<config_path:"/etc/istio/proxy" binary_path:"/usr/local/bin/envoy" service_cluster:"istio-proxy" drain_duration:<seconds:45 > parent_shutdown_duration:<seconds:60 > discovery_address:"istio-pilot.istio-system:15007" discovery_refresh_delay:<seconds:10 > zipkin_address:"zipkin.istio-system:9411" connect_timeout:<seconds:10 > statsd_udp_address:"istio-statsd-prom-bridge.istio-system:9125" proxy_admin_port:15000 > )

2018-06-20T20:36:58.342454Z	info	version root@48d5ddfd72da-docker.io/istio-0.8.0-6f9f420f0c7119ff4fa6a1966a6f6d89b1b4db84-Clean
2018-06-20T20:36:58.342678Z	info	flags (*bootstrap.PilotArgs)(0xc420202180)({
 DiscoveryOptions: (v1.DiscoveryServiceOptions) {
  Port: (int) 8080,
  GrpcAddr: (string) (len=6) ":15010",
  SecureGrpcAddr: (string) (len=6) ":15012",
  MonitoringPort: (int) 9093,
  EnableProfiling: (bool) true,
  EnableCaching: (bool) true,
  WebhookEndpoint: (string) ""
 },
 Namespace: (string) (len=12) "istio-system",
 Mesh: (bootstrap.MeshArgs) {
  ConfigFile: (string) (len=22) "/etc/istio/config/mesh",
  MixerAddress: (string) "",
  RdsRefreshDelay: (*duration.Duration)(<nil>)
 },
 Config: (bootstrap.ConfigArgs) {
  ClusterRegistriesConfigmap: (string) "",
  ClusterRegistriesNamespace: (string) (len=12) "istio-system",
  KubeConfig: (string) "",
  CFConfig: (string) "",
  ControllerOptions: (kube.ControllerOptions) {
   WatchedNamespace: (string) "",
   ResyncPeriod: (time.Duration) 1m0s,
   DomainSuffix: (string) (len=13) "cluster.local"
  },
  FileDir: (string) ""
 },
 Service: (bootstrap.ServiceArgs) {
  Registries: ([]string) (len=1 cap=1) {
   (string) (len=10) "Kubernetes"
  },
  Consul: (bootstrap.ConsulArgs) {
   Config: (string) "",
   ServerURL: (string) "",
   Interval: (time.Duration) 2s
  },
  Eureka: (bootstrap.EurekaArgs) {
   ServerURL: (string) "",
   Interval: (time.Duration) 2s
  }
 },
 RDSv2: (bool) false,
 MeshConfig: (*v1alpha1.MeshConfig)(<nil>),
 CtrlZOptions: (*ctrlz.Options)(0xc42043d560)({
  Port: (uint16) 9876,
  Address: (string) (len=9) "127.0.0.1"
 })
})

2018-06-20T20:36:58.343067Z	warn	Neither --kubeconfig nor --master was specified.  Using the inClusterConfig.  This might not work.
2018-06-20T20:36:58.343610Z	warn	Neither --kubeconfig nor --master was specified.  Using the inClusterConfig.  This might not work.
2018-06-20T20:36:58.344145Z	warn	Neither --kubeconfig nor --master was specified.  Using the inClusterConfig.  This might not work.
2018-06-20T20:36:58.347284Z	info	registering for apiVersion config.istio.io/v1alpha2
2018-06-20T20:36:58.394859Z	info	registering for apiVersion networking.istio.io/v1alpha3
2018-06-20T20:36:58.405580Z	info	registering for apiVersion authentication.istio.io/v1alpha1
2018-06-20T20:36:58.407581Z	info	CRD controller watching namespaces ""
2018-06-20T20:36:58.407915Z	info	Adding Kubernetes registry adapter
2018-06-20T20:36:58.407926Z	info	Primary Cluster name: Kubernetes
2018-06-20T20:36:58.407947Z	info	Service controller watching namespace "" for service, endpoint, nodes and pods, refresh 60000000000
2018-06-20T20:36:58.407986Z	info	Ingress controller watching namespaces ""
2018-06-20T20:36:58.408127Z	info	ingress status syncer publishService istio-system/istio-ingress
2018-06-20T20:36:58.425783Z	info	Setting up event handlers
2018-06-20T20:36:58.426071Z	info	Discovery service started at http=[::]:8080 grpc=[::]:15010
2018-06-20T20:36:58.426372Z	info	Starting Secrets controller
2018-06-20T20:36:58.426450Z	info	Waiting for informer caches to sync
2018-06-20T20:36:58.431050Z	info	ControlZ available at 10.40.1.12:9876
2018-06-20T20:36:58.434608Z	info	Starting GRPC secure on [::]:15012 with certs in /etc/certs/
2018-06-20T20:36:58.442043Z	info	attempting to acquire leader lease...
2018-06-20T20:36:58.452725Z	info	Processing add: istio-system/istio-b
2018-06-20T20:36:58.453100Z	info	Work item failed (waiting till full synchronization), repeating after delay 1s
2018-06-20T20:36:58.503712Z	info	Cleared discovery service cache
2018-06-20T20:36:58.504056Z	info	ads	XDS: Registry event, pushing. Services: 19, VirtualServices: 0, ConnectedEndpoints: 0
2018-06-20T20:36:58.504864Z	info	successfully acquired lease istio-system/istio-ingress-controller-leader-istio
2018-06-20T20:36:58.526605Z	info	Adding new cluster member: istio-b
2018-06-20T20:36:58.528253Z	info	Service controller watching namespace "" for service, endpoint, nodes and pods, refresh 60000000000
2018-06-20T20:36:58.528561Z	info	Number of clusters in the cluster store: 1
2018-06-20T20:36:58.529070Z	error	Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
/workspace/go/src/istio.io/istio/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:72
/workspace/go/src/istio.io/istio/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:65
/workspace/go/src/istio.io/istio/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:51
/usr/local/go/src/runtime/asm_amd64.s:573
/usr/local/go/src/runtime/panic.go:502
/usr/local/go/src/runtime/panic.go:63
/usr/local/go/src/runtime/signal_unix.go:388
/workspace/go/src/istio.io/istio/pilot/pkg/serviceregistry/kube/controller.go:131
/workspace/go/src/istio.io/istio/vendor/k8s.io/client-go/tools/pager/pager.go:39
/workspace/go/src/istio.io/istio/vendor/k8s.io/client-go/tools/pager/pager.go:77
/workspace/go/src/istio.io/istio/vendor/k8s.io/client-go/tools/cache/listwatch.go:97
/workspace/go/src/istio.io/istio/vendor/k8s.io/client-go/tools/cache/reflector.go:249
/workspace/go/src/istio.io/istio/vendor/k8s.io/client-go/tools/cache/reflector.go:204
/workspace/go/src/istio.io/istio/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133
/workspace/go/src/istio.io/istio/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:134
/workspace/go/src/istio.io/istio/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:88
/workspace/go/src/istio.io/istio/vendor/k8s.io/client-go/tools/cache/reflector.go:203
/workspace/go/src/istio.io/istio/vendor/k8s.io/client-go/tools/cache/controller.go:122
/workspace/go/src/istio.io/istio/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:54
/workspace/go/src/istio.io/istio/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:71
/usr/local/go/src/runtime/asm_amd64.s:2361

panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0xd0 pc=0x112aba7]

goroutine 366 [running]:
istio.io/istio/vendor/k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
	/workspace/go/src/istio.io/istio/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:58 +0x107
panic(0x15de700, 0x237c0b0)
	/usr/local/go/src/runtime/panic.go:502 +0x229
istio.io/istio/pilot/pkg/serviceregistry/kube.NewController.func5(0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x17e9327, ...)
	/workspace/go/src/istio.io/istio/pilot/pkg/serviceregistry/kube/controller.go:131 +0x37
istio.io/istio/vendor/k8s.io/client-go/tools/pager.SimplePageFunc.func1(0x1963f80, 0xc420044050, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/workspace/go/src/istio.io/istio/vendor/k8s.io/client-go/tools/pager/pager.go:39 +0x64
istio.io/istio/vendor/k8s.io/client-go/tools/pager.(*ListPager).List(0xc42067f9f8, 0x1963f80, 0xc420044050, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/workspace/go/src/istio.io/istio/vendor/k8s.io/client-go/tools/pager/pager.go:77 +0x105
istio.io/istio/vendor/k8s.io/client-go/tools/cache.(*ListWatch).List(0xc4204fe200, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/workspace/go/src/istio.io/istio/vendor/k8s.io/client-go/tools/cache/listwatch.go:97 +0x147
istio.io/istio/vendor/k8s.io/client-go/tools/cache.(*Reflector).ListAndWatch(0xc420a0fe00, 0xc42083eea0, 0x0, 0x0)
	/workspace/go/src/istio.io/istio/vendor/k8s.io/client-go/tools/cache/reflector.go:249 +0x208
istio.io/istio/vendor/k8s.io/client-go/tools/cache.(*Reflector).Run.func1()
	/workspace/go/src/istio.io/istio/vendor/k8s.io/client-go/tools/cache/reflector.go:204 +0x33
istio.io/istio/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1(0xc420911718)
	/workspace/go/src/istio.io/istio/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133 +0x54
istio.io/istio/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc42067ff18, 0x3b9aca00, 0x0, 0x1, 0xc42083eea0)
	/workspace/go/src/istio.io/istio/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:134 +0xbd
istio.io/istio/vendor/k8s.io/apimachinery/pkg/util/wait.Until(0xc420911718, 0x3b9aca00, 0xc42083eea0)
	/workspace/go/src/istio.io/istio/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:88 +0x4d
istio.io/istio/vendor/k8s.io/client-go/tools/cache.(*Reflector).Run(0xc420a0fe00, 0xc42083eea0)
	/workspace/go/src/istio.io/istio/vendor/k8s.io/client-go/tools/cache/reflector.go:203 +0x157
istio.io/istio/vendor/k8s.io/client-go/tools/cache.(*Reflector).Run-fm(0xc42083eea0)
	/workspace/go/src/istio.io/istio/vendor/k8s.io/client-go/tools/cache/controller.go:122 +0x34
istio.io/istio/vendor/k8s.io/apimachinery/pkg/util/wait.(*Group).StartWithChannel.func1()
	/workspace/go/src/istio.io/istio/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:54 +0x31
istio.io/istio/vendor/k8s.io/apimachinery/pkg/util/wait.(*Group).Start.func1(0xc42050f2e0, 0xc4204fe420)
	/workspace/go/src/istio.io/istio/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:71 +0x4f
created by istio.io/istio/vendor/k8s.io/apimachinery/pkg/util/wait.(*Group).Start
	/workspace/go/src/istio.io/istio/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:69 +0x62
```

"istio-b" is what I named the remote cluster, as instructed by https://istio.io/docs/setup/kubernetes/multicluster-install/#gather-credential-files-from-remote.

**Expected behavior**
I expected the pilot pod to run without crashing.

**Steps to reproduce the bug**
First, create two GKE clusters:

```
$ gcloud container clusters create istio --cluster-version=1.9.7-gke.3 --zone=us-central1-a
$ gcloud container clusters create istio --cluster-version=1.9.7-gke.3 --zone=us-central1-b
```

Then, follow the steps in https://istio.io/docs/setup/kubernetes/multicluster-install/, installing istio in the cluster in us-central1-a and using the cluster in us-central1-b as the remote. I believe I've followed the instructions exactly, although for what it's worth I was getting a different segfault when I accidentally created a secret for the primary cluster (the one in us-central1-a). If you're insterested in that one, I've copied the logs to https://gist.github.com/a-robinson/eaac77f072e5e0a5ef23940e3eb09239.

**Version**
```
$ istioctl version
Version: 0.8.0
GitRevision: 6f9f420f0c7119ff4fa6a1966a6f6d89b1b4db84
User: root@48d5ddfd72da
Hub: docker.io/istio
GolangVersion: go1.10.1
BuildStatus: Clean
$ kubectl version
Client Version: version.Info{Major:"1", Minor:"9", GitVersion:"v1.9.7", GitCommit:"dd5e1a2978fd0b97d9b78e1564398aeea7e7fe92", GitTreeState:"clean", BuildDate:"2018-04-19T00:05:56Z", GoVersion:"go1.9.3", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"9+", GitVersion:"v1.9.7-gke.3", GitCommit:"9b5b719c5f295c99de68ffb5b63101b0e0175376", GitTreeState:"clean", BuildDate:"2018-05-31T18:32:23Z", GoVersion:"go1.9.3b4", Compiler:"gc", Platform:"linux/amd64"}
```

**Is Istio Auth enabled or not?**
I don't know. I ran `kubectl apply -f install/kubernetes/istio-demo.yaml` from the 0.8 release package, as instructed by https://istio.io/docs/setup/kubernetes/quick-start/#installation-steps, which is linked from the steps in https://istio.io/docs/setup/kubernetes/multicluster-install/.

**Environment**
GKE version 1.9.7-gke.3 on the default OS.

--- a/pilot/pkg/config/clusterregistry/conversion.go
+++ b/pilot/pkg/config/clusterregistry/conversion.go
@@ -43,14 +43,10 @@ const (
 	ClusterAccessConfigSecretNamespace = "config.istio.io/accessConfigSecretNamespace"
 )
 
-// Metadata defines a struct used as a key
-type Metadata struct {
-	Name, Namespace string
-}
-
 // RemoteCluster defines cluster struct
 type RemoteCluster struct {
 	Cluster        *k8s_cr.Cluster
+	FromSecret     string
 	Client         *clientcmdapi.Config
 	ClusterStatus  string
 	Controller     *kube.Controller
@@ -59,13 +55,13 @@ type RemoteCluster struct {
 
 // ClusterStore is a collection of clusters
 type ClusterStore struct {
-	rc        map[Metadata]*RemoteCluster
+	rc        map[string]*RemoteCluster
 	storeLock sync.RWMutex
 }
 
 // NewClustersStore initializes data struct to store clusters information
 func NewClustersStore() *ClusterStore {
-	rc := make(map[Metadata]*RemoteCluster)
+	rc := make(map[string]*RemoteCluster)
 	return &ClusterStore{
 		rc: rc,
 	}
@@ -81,8 +77,7 @@ func (cs *ClusterStore) GetClusterAccessConfig(cluster *k8s_cr.Cluster) *clientc
 	if cluster == nil {
 		return nil
 	}
-	key := Metadata{Name: cluster.ObjectMeta.Name, Namespace: cluster.ObjectMeta.Namespace}
-	clusterAccessConfig := cs.rc[key].Client
+	clusterAccessConfig := cs.rc[cluster.ObjectMeta.Name].Client
 	return clusterAccessConfig
 }
 
@@ -155,10 +150,9 @@ func getClustersConfigs(k8s kubernetes.Interface, configMapName, configMapNamesp
 				secretName, secretNamespace, key, err))
 			continue
 		}
-		s := Metadata{Name: cluster.ObjectMeta.Name, Namespace: cluster.ObjectMeta.Namespace}
-		cs.rc[s] = &RemoteCluster{}
-		cs.rc[s].Client = clientConfig
-		cs.rc[s].Cluster = &cluster
+		cs.rc[cluster.ObjectMeta.Name] = &RemoteCluster{}
+		cs.rc[cluster.ObjectMeta.Name].Client = clientConfig
+		cs.rc[cluster.ObjectMeta.Name].Cluster = &cluster
 	}
 
 	return

--- a/pilot/pkg/config/clusterregistry/conversion_test.go
+++ b/pilot/pkg/config/clusterregistry/conversion_test.go
@@ -92,10 +92,8 @@ func TestGetPilotClusters(t *testing.T) {
 		{
 			testName: "3 out of 3 Pilot in the store",
 			cs: &ClusterStore{
-				rc: map[Metadata]*RemoteCluster{
-					{
-						Name:      "cluster1",
-						Namespace: "istio-testing"}: {
+				rc: map[string]*RemoteCluster{
+					"cluster1": {
 						Cluster: &k8s_cr.Cluster{
 							ObjectMeta: metav1.ObjectMeta{
 								Name: "fakePilot1",
@@ -103,9 +101,7 @@ func TestGetPilotClusters(t *testing.T) {
 						},
 						Client: &clientcmdapi.Config{},
 					},
-					{
-						Name:      "cluster2",
-						Namespace: "istio-testing"}: {
+					"cluster2": {
 						Cluster: &k8s_cr.Cluster{
 							ObjectMeta: metav1.ObjectMeta{
 								Name: "fakePilot2",
@@ -113,9 +109,7 @@ func TestGetPilotClusters(t *testing.T) {
 						},
 						Client: &clientcmdapi.Config{},
 					},
-					{
-						Name:      "cluster3",
-						Namespace: "istio-testing"}: {
+					"cluster3": {
 						Cluster: &k8s_cr.Cluster{
 							ObjectMeta: metav1.ObjectMeta{
 								Name: "fakePilot3",

--- a/pilot/pkg/config/clusterregistry/secretcontroller.go
+++ b/pilot/pkg/config/clusterregistry/secretcontroller.go
@@ -16,7 +16,6 @@ package clusterregistry
 
 import (
 	"fmt"
-	"strings"
 	"time"
 
 	corev1 "k8s.io/api/core/v1"
@@ -29,7 +28,6 @@ import (
 	"k8s.io/client-go/tools/cache"
 	"k8s.io/client-go/tools/clientcmd"
 	"k8s.io/client-go/util/workqueue"
-	k8s_cr "k8s.io/cluster-registry/pkg/apis/clusterregistry/v1alpha1"
 
 	"istio.io/istio/pilot/pkg/model"
 	envoy "istio.io/istio/pilot/pkg/proxy/envoy/v2"
@@ -167,119 +165,105 @@ func (c *Controller) runWorker() {
 }
 
 func (c *Controller) processNextItem() bool {
-	key, quit := c.queue.Get()
+	secretName, quit := c.queue.Get()
 
 	if quit {
 		return false
 	}
-	defer c.queue.Done(key)
+	defer c.queue.Done(secretName)
 
-	err := c.processItem(key.(string))
+	err := c.processItem(secretName.(string))
 	if err == nil {
 		// No error, reset the ratelimit counters
-		c.queue.Forget(key)
-	} else if c.queue.NumRequeues(key) < maxRetries {
-		log.Errorf("Error processing %s (will retry): %v", key, err)
-		c.queue.AddRateLimited(key)
+		c.queue.Forget(secretName)
+	} else if c.queue.NumRequeues(secretName) < maxRetries {
+		log.Errorf("Error processing %s (will retry): %v", secretName, err)
+		c.queue.AddRateLimited(secretName)
 	} else {
-		log.Errorf("Error processing %s (giving up): %v", key, err)
-		c.queue.Forget(key)
+		log.Errorf("Error processing %s (giving up): %v", secretName, err)
+		c.queue.Forget(secretName)
 		utilruntime.HandleError(err)
 	}
 
 	return true
 }
 
-func (c *Controller) processItem(key string) error {
-	obj, exists, err := c.informer.GetIndexer().GetByKey(key)
+func (c *Controller) processItem(secretName string) error {
+	obj, exists, err := c.informer.GetIndexer().GetByKey(secretName)
 	if err != nil {
-		return fmt.Errorf("error fetching object with key %s from store: %v", key, err)
+		return fmt.Errorf("error fetching object %s from store: %v", secretName, err)
 	}
 
 	if !exists {
-		c.secretDelete(key)
+		c.deleteMemberCluster(secretName)
 		return nil
 	}
-	c.secretAdd(obj)
+	c.addMemberCluster(secretName, obj.(*corev1.Secret))
 
 	return nil
 }
 
-func addMemberCluster(s *corev1.Secret, c *Controller) {
+func (c *Controller) addMemberCluster(secretName string, s *corev1.Secret) {
 	c.cs.storeLock.Lock()
 	defer c.cs.storeLock.Unlock()
 	// Check if there is already a cluster member with the specified
-	key := Metadata{Name: s.ObjectMeta.Name, Namespace: s.ObjectMeta.Namespace}
-	if _, ok := c.cs.rc[key]; !ok {
-		log.Infof("Adding new cluster member: %s", s.ObjectMeta.Name)
-		clientConfig, err := clientcmd.Load(s.Data[s.ObjectMeta.Name])
-		if err != nil {
-			log.Errorf("failed to load client config from secret %s in namespace %s with error: %v",
-				s.ObjectMeta.Name, s.ObjectMeta.Namespace, err)
-		}
-		cluster := k8s_cr.Cluster{
-			TypeMeta: meta_v1.TypeMeta{
-				Kind:       "Cluster",
-				APIVersion: "clusterregistry.k8s.io/v1alpha1",
-			},
-			ObjectMeta: meta_v1.ObjectMeta{
-				Name:      s.ObjectMeta.Name,
-				Namespace: s.ObjectMeta.Namespace,
-			},
-		}
-		c.cs.rc[key] = &RemoteCluster{}
-		c.cs.rc[key].Client = clientConfig
-		c.cs.rc[key].Cluster = &cluster
-		client, _ := kube.CreateInterfaceFromClusterConfig(clientConfig)
-		kubectl := kube.NewController(client, kube.ControllerOptions{
-			WatchedNamespace: c.watchedNamespace,
-			ResyncPeriod:     c.resyncInterval,
-			DomainSuffix:     c.domainSufix,
-		})
-		c.cs.rc[key].Controller = kubectl
-		c.serviceController.AddRegistry(
-			aggregate.Registry{
-				Name:             serviceregistry.KubernetesRegistry,
-				ClusterID:        GetClusterID(&cluster),
-				ServiceDiscovery: kubectl,
-				ServiceAccounts:  kubectl,
-				Controller:       kubectl,
+	for clusterID, kubeConfig := range s.Data {
+		if _, ok := c.cs.rc[clusterID]; !ok {
+			clientConfig, err := clientcmd.Load(kubeConfig)
+			if err != nil {
+				log.Infof("Data %s in the secret %s in namespace %s is not a kubeconfig: %v",
+					clusterID, secretName, s.ObjectMeta.Namespace, err)
+				continue
+			}
+			log.Infof("Adding new cluster member: %s", clusterID)
+			c.cs.rc[clusterID] = &RemoteCluster{}
+			c.cs.rc[clusterID].Client = clientConfig
+			c.cs.rc[clusterID].FromSecret = secretName
+			client, _ := kube.CreateInterfaceFromClusterConfig(clientConfig)
+			kubectl := kube.NewController(client, kube.ControllerOptions{
+				WatchedNamespace: c.watchedNamespace,
+				ResyncPeriod:     c.resyncInterval,
+				DomainSuffix:     c.domainSufix,
 			})
-		stopCh := make(chan struct{})
-		c.cs.rc[key].ControlChannel = stopCh
-		_ = kubectl.AppendServiceHandler(func(*model.Service, model.Event) { c.discoveryServer.ClearCacheFunc() })
-		_ = kubectl.AppendInstanceHandler(func(*model.ServiceInstance, model.Event) { c.discoveryServer.ClearCacheFunc() })
-
-		go kubectl.Run(stopCh)
+			c.cs.rc[clusterID].Controller = kubectl
+			c.serviceController.AddRegistry(
+				aggregate.Registry{
+					Name:             serviceregistry.KubernetesRegistry,
+					ClusterID:        clusterID,
+					ServiceDiscovery: kubectl,
+					ServiceAccounts:  kubectl,
+					Controller:       kubectl,
+				})
+			stopCh := make(chan struct{})
+			c.cs.rc[clusterID].ControlChannel = stopCh
+			_ = kubectl.AppendServiceHandler(func(*model.Service, model.Event) { c.discoveryServer.ClearCacheFunc() })
+			_ = kubectl.AppendInstanceHandler(func(*model.ServiceInstance, model.Event) { c.discoveryServer.ClearCacheFunc() })
+			go kubectl.Run(stopCh)
+		} else {
+			log.Infof("Cluster %s in the secret %s in namespace %s already exists",
+				clusterID, secretName, s.ObjectMeta.Namespace)
+		}
 	}
 	// TODO Add exporting a number of cluster to Prometheus
 	// for now for debbuging purposes, print it to the log.
 	log.Infof("Number of clusters in the cluster store: %d", len(c.cs.rc))
 }
 
-func deleteMemberCluster(s Metadata, c *Controller) {
+func (c *Controller) deleteMemberCluster(secretName string) {
 	c.cs.storeLock.Lock()
 	defer c.cs.storeLock.Unlock()
 	// Check if there is a cluster member with the specified name
-	if _, ok := c.cs.rc[s]; ok {
-		log.Infof("Deleting cluster member: %s", s)
-		// Deleting Service registry associated with controller
-		c.serviceController.DeleteRegistry(GetClusterID(c.cs.rc[s].Cluster))
-		// Stop controller
-		close(c.cs.rc[s].ControlChannel)
-		<-c.cs.rc[s].ControlChannel
-		// Deleting remote cluster entry from clusters store
-		delete(c.cs.rc, s)
+	for clusterID, cluster := range c.cs.rc {
+		if cluster.FromSecret == secretName {
+			log.Infof("Deleting cluster member: %s", clusterID)
+			// Deleting Service registry associated with controller
+			c.serviceController.DeleteRegistry(clusterID)
+			// Stop controller
+			close(c.cs.rc[clusterID].ControlChannel)
+			<-c.cs.rc[clusterID].ControlChannel
+			// Deleting remote cluster entry from clusters store
+			delete(c.cs.rc, clusterID)
+		}
 	}
 	log.Infof("Number of clusters in the cluster store: %d", len(c.cs.rc))
-}
-
-func (c *Controller) secretAdd(obj interface{}) {
-	s := obj.(*corev1.Secret)
-	addMemberCluster(s, c)
-}
-
-func (c *Controller) secretDelete(key string) {
-	s := Metadata{Name: strings.Split(key, "/")[1], Namespace: strings.Split(key, "/")[0]}
-	deleteMemberCluster(s, c)
 }

--- a/pilot/pkg/config/clusterregistry/secretcontroller.go
+++ b/pilot/pkg/config/clusterregistry/secretcontroller.go
@@ -209,9 +209,15 @@ func (c *Controller) addMemberCluster(secretName string, s *corev1.Secret) {
 	// Check if there is already a cluster member with the specified
 	for clusterID, kubeConfig := range s.Data {
 		if _, ok := c.cs.rc[clusterID]; !ok {
+			// Disregard it if it's empty
+			if len(kubeConfig) == 0 {
+				log.Infof("Data '%s' in the secret %s in namespace %s is empty, and disregarded ",
+					clusterID, secretName, s.ObjectMeta.Namespace)
+				continue
+			}
 			clientConfig, err := clientcmd.Load(kubeConfig)
 			if err != nil {
-				log.Infof("Data %s in the secret %s in namespace %s is not a kubeconfig: %v",
+				log.Infof("Data '%s' in the secret %s in namespace %s is not a kubeconfig: %v",
 					clusterID, secretName, s.ObjectMeta.Namespace, err)
 				continue
 			}

